// reactxp-video/dist/windows/Video.js
"use strict";
/*
* Video.tsx
*
* Copyright (c) Microsoft Corporation. All rights reserved.
* Licensed under the MIT license.
*
* Windows-specific implementation of the cross-platform Video abstraction.
*/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var RN = require("react-native");
var RX = require("reactxp");
var _styles = RN.StyleSheet.create({
    bg: {
        backgroundColor: 'red'
    }
});
var Video = /** @class */ (function (_super) {
    __extends(Video, _super);
    function Video() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // TODO(uwp): #694149 Not implemented
    Video.prototype.render = function () {
        return (React.createElement(RN.View, { style: [this.props.style, _styles.bg] },
            React.createElement(RN.Text, null, "Video")));
    };
    Video.prototype.seek = function (position) {
        // TODO(uwp): #694149 Not implemented
    };
    Video.prototype.seekPercent = function (percentage) {
        // TODO(uwp): #694149 Not implemented
    };
    Video.prototype.play = function () {
        // TODO(uwp): #694149 Not implemented
    };
    Video.prototype.pause = function () {
        // TODO(uwp): #694149 Not implemented
    };
    Video.prototype.stop = function () {
        this.pause();
        this.seek(0);
    };
    Video.prototype.mute = function (muted) {
        // TODO(uwp): #694149 Not implemented
    };
    return Video;
}(RX.Component));
exports.default = Video;
